notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
+        svc = k8s.read_namespaced_service(svc_name, 'default')
 
+        if svc.metadata.labels['user_id'] != escp_user_id:
+            return forbidden()
 
-@app.route('/admin-login', methods=['GET'])
-def admin_login():
-    return render_template('admin-login.html',
-                           form_action_url=external_url_for('admin-login'))
+        del_svc(svc_name)
 
+        pods = k8s.list_namespaced_pod(
+            namespace='default',
+            label_selector=f"uuid={svc.metadata.labels['uuid']}",
+            timeout_seconds=KUBERNETES_TIMEOUT_IN_SECONDS).items
 
-@app.route('/admin-login', methods=['POST'])
-def admin_login_post():
-    if request.form['password'] != ADMIN_PASSWORD:
-        return '403 Forbidden', 403
-    session['admin'] = True
-    return redirect(external_url_for('workers'))
+        if len(pods) == 0:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod not found for given uuid")
+            return '', 200
 
+        if len(pods) > 1:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod_uuid is not unique")
 
-@app.route('/worker-image')
-def worker_image():
-    return '\n'.join(WORKER_IMAGES.values()), 200
+        for pod in pods:
+            if pod.metadata.labels['user_id'] != escp_user_id:
+                return forbidden()
+            del_pod(pod.metadata.name)
 
+        return '', 200
 
-@sockets.route('/wait')
-def wait_websocket(ws):
-    pod_name = session['pod_name']
-    svc_name = session['svc_name']
-    jupyter_token = session['jupyter_token']
-    log.info(f'received wait websocket for {svc_name} {pod_name}')
-    # wait for instance ready
-    while True:
-        try:
-            response = requests.head(f'https://notebook.hail.is/instance-ready/{svc_name}/',
-                                     timeout=1)
-            if response.status_code < 500:
-                log.info(f'HEAD on jupyter succeeded for {svc_name} {pod_name} response: {response}')
-                # if someone responds with a 2xx, 3xx, or 4xx, the notebook
-                # server is alive and functioning properly (in particular, our
-                # HEAD request will return 405 METHOD NOT ALLOWED)
-                break
-            else:
-                # somewhat unusual, means the gateway had an error before we
-                # timed out, usually means the gateway itself is broken
-                log.info(f'HEAD on jupyter failed for {svc_name} {pod_name} response: {response}')
-                gevent.sleep(1)
-            break
-        except requests.exceptions.Timeout as e:
-            log.info(f'GET on jupyter failed for {svc_name} {pod_name}')
-            gevent.sleep(1)
-    ws.send(external_url_for(f'instance/{svc_name}/?token={jupyter_token}'))
-    log.info(f'notification sent to user for {svc_name} {pod_name}')
+    except kube.client.rest.ApiException:
```

Comment:
this except needs to be tightly wrapping the operations in question. If the service is already deleted, we should return a 403 FORBIDDEN (since we cannot confirm if it is owned by the user).

If the pod is already deleted (but the service was not), hmm. Maybe 200 OK? Concurrent deletes could cause this.

Reply:
Good point about ownership. Yes, concurrent deletes is what I'm trying to address.

The long-term way of handling this is to have a non-kubernetes record of user activity. Then you can verify ownership, and make state transitions atomic from the perspective of the user.

Reply:
It should be a 404 whenever the user asks for something they don’t possess

Reply:
https://stackoverflow.com/questions/28582830/access-denied-403-or-404